File: serenity-app/src/test/java/us/nineworlds/serenity/core/ServerConfigTest.java
/**
 * The MIT License (MIT)
 * Copyright (c) 2012 David Carver
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
 * OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package us.nineworlds.serenity.core;

import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.preference.PreferenceManager;

import androidx.test.core.app.ApplicationProvider;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import toothpick.config.Module;
import us.nineworlds.serenity.BuildConfig;
import us.nineworlds.serenity.TestingModule;
import us.nineworlds.serenity.test.InjectingTest;

import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.robolectric.RuntimeEnvironment.application;

@RunWith(RobolectricTestRunner.class)
public class ServerConfigTest extends InjectingTest {

	SharedPreferences.OnSharedPreferenceChangeListener serverConfigChangeListener;
	ServerConfig serverConfig;

	@Before
	public void setUp() throws Exception {
		super.setUp();
		Robolectric.getForegroundThreadScheduler().pause();
		Robolectric.getBackgroundThreadScheduler().pause();
		serverConfig = (ServerConfig) ServerConfig
				.getInstance(getApplicationContext());
		serverConfigChangeListener = serverConfig
				.getServerConfigChangeListener();
	}

	@After
	public void tearDown() {

	}

	@Test
	public void serverPortNewValueIsRetrieved() {
		SharedPreferences prefs = mock(SharedPreferences.class);
		when(prefs.getString("serverport", "32400")).thenReturn("9999");

		serverConfigChangeListener.onSharedPreferenceChanged(prefs,
				"serverport");
		verify(prefs).getString("serverport", "32400");
	}

	@Test
	public void serverHostNewValueIsRetrieved() {
		SharedPreferences prefs = mock(SharedPreferences.class);
		when(prefs.getString("server", "")).thenReturn("10.0.0.3");

		serverConfigChangeListener.onSharedPreferenceChanged(prefs, "server");
		verify(prefs).getString("server", "");
	}

	@Test
	public void serverSetsNewServerAddressBasedOnDiscoveredServers() {
		Editor reditor = PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).edit();

		SharedPreferences prefs = mock(SharedPreferences.class);
		when(prefs.getString("discoveredServer", "")).thenReturn("10.0.0.3");
		when(prefs.edit()).thenReturn(reditor);

		serverConfigChangeListener.onSharedPreferenceChanged(prefs,
				"discoveredServer");
		verify(prefs).getString("discoveredServer", "");
	}

	@Test
	public void setHostSetsCorrectValue() {
		serverConfig.setHost("10.0.0.4");
		assertThat(serverConfig.getHost()).isEqualTo("10.0.0.4");
	}

	@Test
	public void setPortSetsExpectedValue() {
		serverConfig.setPort("6666");
		assertThat(serverConfig.getPort()).isEqualTo("6666");
	}

	@Override
	public void installTestModules() {
		scope.installTestModules(new TestingModule(), new TestModule());
	}

	public class TestModule extends Module {
		public TestModule() {
			bind(SharedPreferences.class).toInstance(PreferenceManager.getDefaultSharedPreferences(ApplicationProvider.getApplicationContext()));
		}
	}
}

File: serenity-app/src/test/kotlin/us/nineworlds/serenity/core/ServerConfigTest.kt
package us.nineworlds.serenity.core

import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import android.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.core.app.ApplicationProvider.getApplicationContext
import assertk.assertThat
import assertk.assertions.isEqualTo
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.LooperMode
import toothpick.config.Module
import us.nineworlds.serenity.MockkTestingModule
import us.nineworlds.serenity.test.InjectingTest

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.LEGACY)
class ServerConfigTest : InjectingTest() {

  private lateinit var serverConfigChangeListener: SharedPreferences.OnSharedPreferenceChangeListener
  private lateinit var serverConfig: ServerConfig

  private val mockPrefs: SharedPreferences = mockk(relaxed = true)

  @Before
  override fun setUp() {
    super.setUp()
    Robolectric.getForegroundThreadScheduler().pause()
    Robolectric.getBackgroundThreadScheduler().pause()
    serverConfig = ServerConfig.getInstance(getApplicationContext()) as ServerConfig
    serverConfigChangeListener = serverConfig.serverConfigChangeListener
  }

  @After
  fun tearDown() {
    clearAllMocks()
  }

  @Test
  fun serverPortNewValueIsRetrieved() {
    every { mockPrefs.getString("serverport", "32400") } returns "9999"

    serverConfigChangeListener.onSharedPreferenceChanged(mockPrefs, "serverport")

    verify { mockPrefs.getString("serverport", "32400") }
  }

  @Test
  fun serverHostNewValueIsRetrieved() {
    every { mockPrefs.getString("server", "") } returns "10.0.0.3"

    serverConfigChangeListener.onSharedPreferenceChanged(mockPrefs, "server")

    verify { mockPrefs.getString("server", "") }
  }

  @Test
  fun serverSetsNewServerAddressBasedOnDiscoveredServers() {
    val reditor: Editor = PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).edit()
    every { mockPrefs.getString("discoveredServer", "") } returns "10.0.0.3"
    every { mockPrefs.edit() } returns reditor

    serverConfigChangeListener.onSharedPreferenceChanged(mockPrefs, "discoveredServer")

    verify { mockPrefs.getString("discoveredServer", "") }
  }

  @Test
  fun setHostSetsCorrectValue() {
    serverConfig.host = "10.0.0.4"
    assertThat(serverConfig.host).isEqualTo("10.0.0.4")
  }

  @Test
  fun setPortSetsExpectedValue() {
    serverConfig.port = "6666"
    assertThat(serverConfig.port).isEqualTo("6666")
  }

  override fun installTestModules() {
    scope.installTestModules(MockkTestingModule(), TestModule())
  }

  inner class TestModule : Module() {
    init {
      bind(SharedPreferences::class.java).toInstance(
        PreferenceManager.getDefaultSharedPreferences(
          ApplicationProvider.getApplicationContext()
        )
      )
    }
  }
}
